### PR TITLE
(Update on PR #543) Fixed TypeError in `LU_comp`, updated `determinant` docs and added `rank` function for matrices

### PR DESCRIPTION
**Pull request by jan-philipp-hoffmann**
This is an update on PR #543 

- fixed an issue in `LU_comp` which raised TypeError on singular matrices
- added documentation of `determinant` function of matrices 
- added `rank` function of matrices 

(PR #543 was originally opened and closed by _sonntagsgesicht_)

### CODE REVIEW

**Comment by skirpichev** [comment]
There are test failures.

**Comment by jan-philipp-hoffmann** [comment]
minor changes to fixe test failures. tests should work now (as you can check [here](https://github.com/sonntagsgesicht/mpmath/actions/runs/4948074926/jobs/8848372124)).

File: mpmath/matrices/linalg.py [hunk 1]
```diff
@@ -787,3 +839,58 @@ def qr(ctx, A, mode = 'full', edps = 10):
...
+            i.e. has an inverse matrix.
+
+            >>> B = matrix([[3, -1, 1],[2, 1, 0],[-2, 1, -1]])
+            >>> A*B == eye(3)
+            True
+            >>> rank(B)
+            3
+
+        """
+        if iszerofunc is None:
```
Comment (skirpichev) [inline mpmath/matrices/linalg.py:894]: Only iszerofunc=None was tested, see https://app.codecov.io/gh/mpmath/mpmath/pull/610.  Could you add missing tests?

File: mpmath/matrices/linalg.py [hunk 2]
```diff
@@ -787,3 +839,58 @@ def qr(ctx, A, mode = 'full', edps = 10):
         # ------------------
         # END OF FUNCTION QR
         # ------------------
+
+    def rank(ctx, A, iszerofunc=None):
+        """
+            Calculate the rank of a matrix,
```
Comment (skirpichev) [inline mpmath/matrices/linalg.py:845]: Why this extra indentation in the docstring?

File: mpmath/matrices/linalg.py [hunk 3]
```diff
@@ -787,3 +839,58 @@ def qr(ctx, A, mode = 'full', edps = 10):
...
+            columns (or rows equivalently).
+
+            Rank is computed via singular value decomposition
+            by counting the number of non-zero singular values.
+
+            **Examples**
+
+            Rank of identity is same as its dimension.
+
+            >>> from mpmath import *
```
Comment (skirpichev) [inline mpmath/matrices/linalg.py:856]: Please avoid star imports

File: mpmath/matrices/linalg.py [hunk 4]
```diff
@@ -533,7 +536,56 @@ def cholesky_solve(ctx, A, b, **kwargs):
...
+        Calculate the determinant of a square matrix.
+        The determinant is the normed, alternating n-linear from,
+        i.e. a multiplicative map for each matrix into the
+        field of numbers of its entries.
+
+        **Examples**
+
+        Determinant of identity is 1.
+
+        >>> from mpmath import *
```
Comment (skirpichev) [inline mpmath/matrices/linalg.py:548]: Please avoid star imports

**Review by skirpichev — CHANGES_REQUESTED** [review]
Looks good, except for minor nitpicks above.

Could you, please, also do a rebase?  Probably, there should be four commits:
* update on .gitignore
* issue in LU_comp (with a test)
* det docs
* rank function (with tests)

File: mpmath/matrices/linalg.py [hunk 5]
```diff
@@ -533,7 +536,56 @@ def cholesky_solve(ctx, A, b, **kwargs):
 
     def det(ctx, A):
         """
-        Calculate the determinant of a matrix.
+        Calculate the determinant of a square matrix.
+        The determinant is the normed, alternating n-linear from,
+        i.e. a multiplicative map for each matrix into the
+        field of numbers of its entries.
```
Comment (vks) [inline mpmath/matrices/linalg.py:542]:
```suggestion
        Calculate the determinant of a square matrix.
        
        The determinant is the normed, alternating n-linear form,
        i.e. a multiplicative map for each matrix into the
        field of numbers of its entries.
```

File: mpmath/matrices/linalg.py [hunk 6]
```diff
@@ -134,6 +134,9 @@ def LU_decomp(ctx, A, overwrite=False, use_cache=True):
                 if current > biggest: # TODO: what if equal?
                     biggest = current
                     p[j] = k
+            # without pivot LU fails
+            if p[j] is None:
```
Comment (vks) [inline mpmath/matrices/linalg.py:138]: Is this what happens for the test case `A11`?

File: mpmath/matrices/linalg.py [hunk 7]
```diff
@@ -787,3 +839,58 @@ def qr(ctx, A, mode = 'full', edps = 10):
...
         # ------------------
+
+    def rank(ctx, A, iszerofunc=None):
+        """
+            Calculate the rank of a matrix,
+            i.e. the number of linear independent
+            columns (or rows equivalently).
+
+            Rank is computed via singular value decomposition
+            by counting the number of non-zero singular values.
```
Comment (vks) [inline mpmath/matrices/linalg.py:850]: Could you please document the `iszerofunc` parameter, and maybe add an example using it?

**Review by vks — CHANGES_REQUESTED** [review]
Looks good from my side, I only had minor additional comments. I agree with @skirpichev's comments.

File: mpmath/matrices/linalg.py (hunk 6 above)
Comment (jan-philipp-hoffmann) [reply mpmath/matrices/linalg.py:138]: yes

File: mpmath/matrices/linalg.py (hunk 1 above)
Comment (jan-philipp-hoffmann) [reply mpmath/matrices/linalg.py]: removed `iszerofunc` as `mpmath` handles precision on its own.

File: mpmath/matrices/linalg.py (hunk 7 above)
Comment (jan-philipp-hoffmann) [reply mpmath/matrices/linalg.py]: added tests some example for `iszerofunc`

File: mpmath/matrices/linalg.py [hunk 8]
```diff
@@ -787,3 +840,77 @@ def qr(ctx, A, mode = 'full', edps = 10):
         # ------------------
         # END OF FUNCTION QR
         # ------------------
+
+    def rank(ctx, A, iszerofunc=None):
+        """
+        Calculate the rank of a matrix,
+        i.e. the number of linear independent
+        columns (or rows equivalently).
```
Comment (skirpichev) [inline mpmath/matrices/linalg.py:848]:
```suggestion
        Calculate the rank of a matrix.
        
        This corresponds to the maximal number of linearly independent
        columns (or rows equivalently).
```

**Review by skirpichev — APPROVED** [review]

**Pull request merged**